server/src/main/java/org/elasticsearch/search/retriever/RankDocsRetrieverBuilder.java

```diff
@@ -124,19 +130,39 @@ public void extractToSearchSourceBuilder(SearchSourceBuilder searchSourceBuilder
             boolQuery.filter(preFilterQueryBuilder);
         }
         searchSourceBuilder.query(boolQuery);
+        searchSourceBuilder.rankBuilder(
+            new RankDocsRankBuilder(rankWindowSize, sources.stream().map(RetrieverBuilder::retrieverName).toList())
+        );
+    }
+
+    private boolean hasAggregations(SearchSourceBuilder searchSourceBuilder) {
+        return searchSourceBuilder.aggregations() != null;
+    }
+
+    private boolean isExplainRequest(SearchSourceBuilder searchSourceBuilder) {
+        return searchSourceBuilder.explain() != null && searchSourceBuilder.explain();
+    }
+
+    private boolean isProfileRequest(SearchSourceBuilder searchSourceBuilder) {
+        return searchSourceBuilder.profile();
+    }
+
+    private boolean shouldTrackTotalHits(SearchSourceBuilder searchSourceBuilder) {
+        return searchSourceBuilder.trackTotalHitsUpTo() != null
+            && searchSourceBuilder.trackTotalHitsUpTo() == SearchContext.TRACK_TOTAL_HITS_ACCURATE;
```

Comment:
We should track total hits whenever it is > 0. We can optimise (disable) if the number of rank docs is already greater than the provided value but otherwise it is needed:
```suggestion
    private boolean shouldTrackTotalHits(SearchSourceBuilder searchSourceBuilder) {
        return searchSourceBuilder.trackTotalHitsUpTo() == null || searchSourceBuilder.trackTotalHitsUpTo() > rankDocs.size();
```

Reply:
Makes sense! IIUC the above snippet though implies that we should also track hits even when `track_total_hits` is omitted? Should we maybe avoid this due to the additional cost of rerunning all parent queries by default?

Reply:
I think we can start simple. The current [`QueryPhase#executeRank`](https://github.com/elastic/elasticsearch/blob/60523315316cc48ed50fed80e96ecfb1385836a1/server/src/main/java/org/elasticsearch/search/query/QueryPhase.java#L74) pays the same price and checking the rank docs size in the new path is already an optimisation.
I am confident we can optimise this path further without changing the default value of `track_total_hits`. Returning total hits up to `10_000` is important for pagination.